src/nix/repl.cc

```diff
@@ -514,6 +516,29 @@ bool NixRepl::processLine(string line)
                 logger->cout("  %s -> %s", outputName, state->store->printStorePath(outputPath));
         } else if (command == ":i") {
             runNix("nix-env", {"-i", drvPathRaw});
+        } else if (command == ":log") {
+            settings.readOnlyMode = true;
+            Finally roModeReset([&]() {
+                settings.readOnlyMode = false;
+            });
+            auto subs = getDefaultSubstituters();
+
+            subs.push_front(state->store);
+
+            bool foundLog = false;
+
+            RunPager pager;
+            for (auto & sub : subs) {
+                auto log = sub->getBuildLog(drvPath);
+                if (!log) continue;
+                stopProgressBar();
+                printInfo("got build log for '%s' from '%s'", drvPathRaw, sub->getUri());
+                std::cout << *log;
+                foundLog = true;
+                break;
+            }
+
+            if (!foundLog) throw Error("build log of '%s' is not available", drvPathRaw);
```

Comment:
I haven't tested this exact code, but something like:
```suggestion
            std::shared_ptr<std::string> log = null;
            for (auto & sub : subs) {
                log = sub->getBuildLog(drvPath);
                if (log) break;
            }
            
            if (log) {
                RunPager pager;
                stopProgressBar();
                printInfo("got build log for '%s' from '%s'", drvPathRaw, sub->getUri());
                std::cout << *log;
            }
            else throw Error("build log of '%s' is not available", drvPathRaw);
```

Reply:
Actually, since we're using the `sub` to print the `got build log` message, this makes it more complicated and messy. I think the original is fine for now.

Reply:
Oh good call. A more minor tweak then maybe:
```suggestion
            bool foundLog = false;

            RunPager pager;
            for (auto & sub : subs) {
                auto log = sub->getBuildLog(drvPath);
                if (log) {
                    stopProgressBar();
                    printInfo("got build log for '%s' from '%s'", drvPathRaw, sub->getUri());
                    std::cout << *log;
                    foundLog = true;
                    break;
                }
            }

            if (!foundLog) throw Error("build log of '%s' is not available", drvPathRaw);
```